### PR TITLE
Add section about extensible variant types.

### LINKED ISSUE #135: Extensible variants are undocumented
transferred from https://github.com/rescript-lang/rescript-compiler/issues/4869 by @earksiinni

I encountered the following code in the wild:

```
type thunk('state) = ..;

type ReduxThunk.thunk(_) +=
  | StringAction (stringAction)
  | CounterAction (action);
```

This code essentially creates an "extensible variant" called `thunk` and adds two constructors to it. I'm not certain whether I'm using the right term to explain this feature since it's not described anywhere in the ReasonML or ReScript docs. It should be.

The ReScript docs also don't explain `..`, which is an elision, which is documented in [ReasonML's docs on open objects](https://reasonml.github.io/docs/en/object). ReScript's docs don't talk about open objects at all, let alone elisions, but both sets of docs omit the fact that you can apparently use elisions to define extensible variants.

The use of the `+=` feature isn't described anywhere in either set of docs, either.

It would be great to describe all these features, and while we're at it it would be great to have ReScript docs on polymorphic variants and how they differ from this kind of extensible non-polymorphic variant.

### PR DESCRIPTION
**Pull request by fhammerschmidt**
Fixes #135

### CODE REVIEW

**Comment by fhammerschmidt** [comment]
Added this under a new sidebar section: "Advanced Language Features".

**Comment by ryyppy** [comment]
we are coming back to this soon. Looking good so far though

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+---
+title: "Extensible Variant"
+description: "Extensible Variants in ReScript"
+canonical: "/docs/manual/latest/extensible-variant"
+---
+
+# Extensible Variant Types
+
+Variants can be made extensible by defining a type using `..` and adding new variant constructors using `+=`.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:9]:
```suggestion
Variant types are usually constraint to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type (e.g. in different places of your program). For this, we offer Extensible Variant types.

## Definition and Usage
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+var Other = Caml_exceptions.create("Playground.Other");
+
+var Point = Caml_exceptions.create("Playground.Point");
+
+var Line = Caml_exceptions.create("Playground.Line");
+```
+
+</CodeTab>
+
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:33]:
```suggestion

The `..` in the type declaration above defines an extensible variant type. The `+=` operator is used to add constructors to an Extensible Variant type. 

```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+Pattern-matching is possible the same way as with normal variants but there is one caveat:
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:34]:
```suggestion
## Pattern Matching Caveats

Extensible Variants are open-ended, so the compiler will not be able to exhaustively pattern match all available cases. You will always need to provide a default `_` case for every `switch` expression. 
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+With extensible variants, the possibility to check for exhaustiveness vanishes, you always need a default case `_` here.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:35]:
```suggestion
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    console.log("Point", [v._1, v._2]);
+  } else if (v.RE_EXN_ID === Line) {
+    console.log("Line", [v._1, v._2, v._3, v._4]);
+  } else {
+    console.log("Other");
+  } 
+}
+```
+
+</CodeTab>
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:62]:
```suggestion
</CodeTab>

## Tips & Tricks

**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as Extensible Variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`.

It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 6]
```diff
@@ -0,0 +1,62 @@
+---
+title: "Extensible Variant"
+description: "Extensible Variants in ReScript"
+canonical: "/docs/manual/latest/extensible-variant"
+---
+
+# Extensible Variant Types
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:7]:
```suggestion
# Extensible Variant
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    console.log("Other");
+  } 
+}
+```
+
+</CodeTab>
+
+## Tips & Tricks
+
+**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as Extensible Variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:71]:
```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    console.log("Other");
+  } 
+}
+```
+
+</CodeTab>
+
+## Tips & Tricks
+
+**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:71]:
```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were extensible variants make sense. We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+## Pattern Matching Caveats
+
+Extensible variants are open-ended, so the compiler will not be able to exhaustively pattern match all available cases. You will always need to provide a default `_` case for every `switch` expression. 
+
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+
+```res example
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:46]:
```suggestion
```res
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 10]
```diff
@@ -0,0 +1,71 @@
+---
+title: "Extensible Variant"
+description: "Extensible Variants in ReScript"
+canonical: "/docs/manual/latest/extensible-variant"
+---
+
+# Extensible Variant
+
+Variant types are usually constraint to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type (e.g. in different places of your program). For this, we offer extensible variant types.
```
Comment (fhammerschmidt) [inline pages/docs/manual/latest/extensible-variant.mdx:9]: are usually **constrained** ...

Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:9]:
```suggestion
Variant types are usually constrained to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type, even after the initial type declaration (e.g. in different places of your program). For this, we offer extensible variant types.
```

Comment (fhammerschmidt) [reply pages/docs/manual/latest/extensible-variant.mdx:9]: Perfect!

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 11]
```diff
@@ -0,0 +1,71 @@
+---
+title: "Extensible Variant"
+description: "Extensible Variants in ReScript"
+canonical: "/docs/manual/latest/extensible-variant"
+---
+
+# Extensible Variant
+
+Variant types are usually constrained to a fixed set of constructors. There may be very rare cases where you still want to be able to add constructors to a variant type (even after its initial type declaration). For this, we offer extensible variant types.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:9]:
```suggestion
Variant types are usually constrained to a fixed set of constructors. There may be very rare cases where you still want to be able to add constructors to a variant type even after its initial type declaration. For this, we offer extensible variant types.
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    console.log("Other");
+  } 
+}
+```
+
+</CodeTab>
+
+## Tips & Tricks
+
+**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were extensible variants make sense. We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:71]:
```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's one of the very few use-case where extensible variants make sense.

We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

File: data/sidebar_manual_latest.json [hunk 13]
```diff
@@ -33,6 +33,9 @@
     "unboxed",
     "reserved-keywords"
   ],
+  "Advanced Language Features": [
```
Comment (ryyppy) [inline data/sidebar_manual_latest.json:36]:
```suggestion
  "Advanced Features": [
```

File: pages/docs/manual/latest/extensible-variant.mdx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+var Other = Caml_exceptions.create("Playground.Other");
+
+var Point = Caml_exceptions.create("Playground.Point");
+
+var Line = Caml_exceptions.create("Playground.Line");
+```
+
+</CodeTab>
+
+The `..` in the type declaration above defines an extensible variant `type t`. The `+=` operator is then used to add constructors to the given type.
```
Comment (ryyppy) [inline pages/docs/manual/latest/extensible-variant.mdx:36]:
```suggestion
The `..` in the type declaration above defines an extensible variant `type t`. The `+=` operator is then used to add constructors to the given type. 

**Note:** Don't forget the leading `type` keyword when using the `+=` operator!
```

**Pull request merged**